vendor/github.com/rancher/norman/controller/generic_controller.go

```diff
@@ -124,26 +124,37 @@ func (g *genericController) Enqueue(namespace, name string) {
 
 func (g *genericController) AddHandler(ctx context.Context, name string, handler HandlerFunc) {
 	g.Lock()
+	defer g.Unlock()
+
+	g.generation++
 	h := &handlerDef{
 		name:       name,
 		generation: g.generation,
 		handler:    handler,
 	}
-	g.handlers = append(g.handlers, h)
-	g.Unlock()
 
-	go func() {
+	go func(gen int) {
 		<-ctx.Done()
 		g.Lock()
-		var handlers []*handlerDef
+		defer g.Unlock()
+		var newHandlers []*handlerDef
 		for _, handler := range g.handlers {
-			if handler != h {
-				handlers = append(handlers, h)
+			if handler.generation == gen {
+				continue
 			}
+			newHandlers = append(newHandlers, handler)
+		}
+		g.handlers = newHandlers
+	}(h.generation)
+
+	g.handlers = append(g.handlers, h)
```

Comment:
Followup based on conversation with @dramich, 
`<- ctx.Done()` is the critical piece here. This code will only run when the calling ctx is canelled. This goroutine essentially functions as _Deregister_ and removes the handler from the controller. 
The code change is required here because we are no longer duplicating the Mangement context (see changes in context.go)  which we were before.
This means that we need to ensure that we are not deleting that points to mgmt context